---
title: "Python"
author: "Cara Yijin Zou"
date: '2019-12-12'
slug: python
categories: []
tags: []
description: ''
---



<div id="k-mers" class="section level2">
<h2>K-mers</h2>
<p>In addition to R, Python is a valuable tool that can be applied to biology. For example, k-mers are useful in genetics and can be used for a variety of reasons such as sequence assembly, taxonomy, manipulating protein expression, etc. Below is a function that is able to find k-mers in a sequence and record the frequency number in a dictionary.</p>
<pre class="python"><code>
my_seq1 = &quot;CAGCCCAATC&quot;
  
def findkmers(k, seq):
    dict1={}
    a = 0 
    b= k
    k_mer= seq[a:b]
    while len(k_mer)==k: 
        if k_mer in dict1: 
            dict1[k_mer]+= 1
        else: 
            dict1[k_mer]= 1
        a+=1
        b+=1
        k_mer= seq[a:b]
    print (dict1)

def main():
   print(&quot;For the sequence: CAGCCCAATC&quot;)
   for i in range (1,5):
      print(&quot;The&quot;, str(i)+&quot;-mers are :&quot;)
      findkmers(i, my_seq1)

main()</code></pre>
<pre><code>## For the sequence: CAGCCCAATC
## The 1-mers are :
## {&#39;C&#39;: 5, &#39;A&#39;: 3, &#39;G&#39;: 1, &#39;T&#39;: 1}
## The 2-mers are :
## {&#39;CA&#39;: 2, &#39;AG&#39;: 1, &#39;GC&#39;: 1, &#39;CC&#39;: 2, &#39;AA&#39;: 1, &#39;AT&#39;: 1, &#39;TC&#39;: 1}
## The 3-mers are :
## {&#39;CAG&#39;: 1, &#39;AGC&#39;: 1, &#39;GCC&#39;: 1, &#39;CCC&#39;: 1, &#39;CCA&#39;: 1, &#39;CAA&#39;: 1, &#39;AAT&#39;: 1, &#39;ATC&#39;: 1}
## The 4-mers are :
## {&#39;CAGC&#39;: 1, &#39;AGCC&#39;: 1, &#39;GCCC&#39;: 1, &#39;CCCA&#39;: 1, &#39;CCAA&#39;: 1, &#39;CAAT&#39;: 1, &#39;AATC&#39;: 1}</code></pre>
</div>
<div id="sequence-alignment" class="section level2">
<h2>Sequence Alignment</h2>
<p>Sequence alignment is a very important technique used in genetics and is useful for both protein and nucleotide sequences. For example, in my research lab, we use BLAST pairwise alignment on a sequence obtained from DNA sequencing to verify that the DNA matches the known sequence or to check if a desired mutation was successful. Both the Needleman-Wunsch algorithm and Smith-Waterman algorithm are useful ways to find the best alignment.</p>
<p>Here is an example of alignment scores between a sequence in a <em>Mycobacterium tuberculosis</em> sequence compared to humans (low alignment scores) and all organisms (high alignment scores).</p>
<p><img src="/BLASThumanMtDala.png" alt="BLAST Nucleotide MtDala against Human genomic plus transcript (Human G+T)" />
<img src="/BLASTMtDalaOther.png" alt="BLAST Nucleotide MtDala against All Organisms" /></p>
<pre class="python"><code># Use these values to calculate scores
match_score = 2
mismatch_score = -1
gap_penalty = -1

# Make a score matrix with these two sequences
seq1 = &quot;ATGAT&quot;
seq2 = &quot;AAGGAT&quot;

def match_fun(a,b):  # Helper function to define match/mismatch scores
    if a==b:
        score=2 #match score
    else:
        score=-1 #mismatch score
    return score
    
def nw_fun(seq1,seq2,gap_penalty=-1, match_score=1, mismatch_score=-1):
    m=len(seq1) #length of horizontal sequence
    n=len(seq2) #length of vertical sequence
    score =[]
    for row in range(n+1):
        r=[]
        #creates one row
        for col in range(m+1):
            r.append(0)
        score.append(r)
   
    for j in range(0, m + 1): score[0][j] = gap_penalty * j
    for i in range(0, n + 1): score[i][0] = gap_penalty * i
    
    for i in range(1, n + 1): 
        for j in range(1, m + 1):
            insert = score[i - 1][j] + gap_penalty 
            delete = score[i][j - 1] + gap_penalty 
            match = score[i - 1][j - 1] + match_score if seq1[j-1]==seq2[i-1] else score[i-1][j-1]+ mismatch_score
            score[i][j] = max(match, delete, insert)
    return score
    
def sw_fun(seq1,seq2,gap_penalty=-1, match_score=1, mismatch_score=-1):
    m=len(seq1) #length of horizontal sequence
    n=len(seq2) #length of vertical sequence
    score =[]
    for row in range(n+1):
        r=[]
        #creates one row
        for col in range(m+1):
            r.append(0)
        score.append(r)
   
    for j in range(0, m + 1): score[0][j] = gap_penalty * j
    for i in range(0, n + 1): score[i][0] = gap_penalty * i
    
    for i in range(1, n + 1): 
        for j in range(1, m + 1):
            insert = score[i - 1][j] + gap_penalty 
            delete = score[i][j - 1] + gap_penalty 
            match = score[i - 1][j - 1] + match_score if seq1[j-1]==seq2[i-1] else mismatch_score
            score[i][j] = max(match, delete, insert)
            
    for i in range(n+1): 
        for j in range(m+1): 
            if score[i][j]&lt;0: 
                score[i][j]=0
    return score

def display(matrix):
    numRows=len(matrix)
    numCols=len(matrix[0]) #should be equal for all, pick any row
    print(format(&quot;-&quot;, &quot;&gt;6s&quot;), end=&quot;&quot;) 
    for ch in seq1: 
        print(format(ch, &quot;&gt;3s&quot;), end=&quot;&quot;)
    print (&quot;&quot;)
    seq2wdash=&quot;-&quot;+seq2
    index=0
    for row in range (numRows):
        print (format(seq2wdash[index],&quot;3s&quot;), end=&quot;&quot;)
        index +=1 
        for col in range(numCols): 
            print(format(matrix[row][col], &quot;3d&quot;), end=&quot;&quot;)
        print(&quot;&quot;)

def main2(): 
    print(&quot;For the two sequences:&quot;, seq1,&quot;and&quot;, seq2)
    print(&quot;Needleman-Wunsch&quot;)
    nwmatrix= nw_fun(seq1,seq2,gap_penalty,match_score,mismatch_score)
    display(nwmatrix)   
    
    print(&quot;&quot;)
    
    print(&quot;Smith-Waterman&quot;)
    swmatrix= sw_fun(seq1,seq2,gap_penalty,match_score,mismatch_score)
    display(swmatrix) 

main2()</code></pre>
<pre><code>## For the two sequences: ATGAT and AAGGAT
## Needleman-Wunsch
##      -  A  T  G  A  T
## -    0 -1 -2 -3 -4 -5
## A   -1  2  1  0 -1 -2
## A   -2  1  1  0  2  1
## G   -3  0  0  3  2  1
## G   -4 -1 -1  2  2  1
## A   -5 -2 -2  1  4  3
## T   -6 -3  0  0  3  6
## 
## Smith-Waterman
##      -  A  T  G  A  T
## -    0  0  0  0  0  0
## A    0  2  1  0  0  0
## A    0  1  0  0  2  1
## G    0  0  0  2  1  0
## G    0  0  0  1  0  0
## A    0  0  0  0  3  2
## T    0  0  0  0  2  5</code></pre>
<p>After scores are calculated, the optimum alignment sequence can be found:</p>
<p><img src="/SeqAlignmentMtDala.png" /></p>
</div>
