---
title: "Python"
author: "Cara Yijin Zou"
date: '2019-12-12'
slug: python
categories: []
tags: []
description: ''
---

## K-mers

In addition to R, Python is a valuable tool that can be applied to biology. For example, k-mers are useful in genetics and can be used for a variety of reasons such as sequence assembly, taxonomy, manipulating protein expression, etc. Below is a function that is able to find k-mers in a sequence and record the frequency number in a dictionary. 

```{python}

my_seq1 = "CAGCCCAATC"
  
def findkmers(k, seq):
    dict1={}
    a = 0 
    b= k
    k_mer= seq[a:b]
    while len(k_mer)==k: 
        if k_mer in dict1: 
            dict1[k_mer]+= 1
        else: 
            dict1[k_mer]= 1
        a+=1
        b+=1
        k_mer= seq[a:b]
    print (dict1)

def main():
   print("For the sequence: CAGCCCAATC")
   for i in range (1,5):
      print("The", str(i)+"-mers are :")
      findkmers(i, my_seq1)

main()
```



## Sequence Alignment 
Sequence alignment is a very important technique used in genetics and is useful for both protein and nucleotide sequences. For example, in my research lab, we use BLAST pairwise alignment on a sequence obtained from DNA sequencing to verify that the DNA matches the known sequence or to check if a desired mutation was successful. Both the Needleman-Wunsch algorithm (global alignment) and Smith-Waterman algorithm (local alignment) are useful ways to find the best alignment. 

Here is an example of alignment scores between a sequence found in the *Mycobacterium tuberculosis* genome compared to humans (low alignment scores as expected) and all organisms (higher alignment scores). 

![BLAST Nucleotide MtDala against Human genomic plus transcript (Human G+T)](/BLASThumanMtDala.png)
![BLAST Nucleotide MtDala against All Organisms](/BLASTMtDalaOther.png)


```{python}
# Use these values to calculate scores
match_score = 2
mismatch_score = -1
gap_penalty = -1

# Make a score matrix with these two sequences
seq1 = "ATGAT"
seq2 = "AAGGAT"

def match_fun(a,b):  # Helper function to define match/mismatch scores
    if a==b:
        score=2 #match score
    else:
        score=-1 #mismatch score
    return score
    
def nw_fun(seq1,seq2,gap_penalty=-1, match_score=1, mismatch_score=-1):
    m=len(seq1) #length of horizontal sequence
    n=len(seq2) #length of vertical sequence
    score =[]
    for row in range(n+1):
        r=[]
        #creates one row
        for col in range(m+1):
            r.append(0)
        score.append(r)
   
    for j in range(0, m + 1): score[0][j] = gap_penalty * j
    for i in range(0, n + 1): score[i][0] = gap_penalty * i
    
    for i in range(1, n + 1): 
        for j in range(1, m + 1):
            insert = score[i - 1][j] + gap_penalty 
            delete = score[i][j - 1] + gap_penalty 
            match = score[i - 1][j - 1] + match_score if seq1[j-1]==seq2[i-1] else score[i-1][j-1]+ mismatch_score
            score[i][j] = max(match, delete, insert)
    return score
    
def sw_fun(seq1,seq2,gap_penalty=-1, match_score=1, mismatch_score=-1):
    m=len(seq1) #length of horizontal sequence
    n=len(seq2) #length of vertical sequence
    score =[]
    for row in range(n+1):
        r=[]
        #creates one row
        for col in range(m+1):
            r.append(0)
        score.append(r)
   
    for j in range(0, m + 1): score[0][j] = gap_penalty * j
    for i in range(0, n + 1): score[i][0] = gap_penalty * i
    
    for i in range(1, n + 1): 
        for j in range(1, m + 1):
            insert = score[i - 1][j] + gap_penalty 
            delete = score[i][j - 1] + gap_penalty 
            match = score[i - 1][j - 1] + match_score if seq1[j-1]==seq2[i-1] else mismatch_score
            score[i][j] = max(match, delete, insert)
            
    for i in range(n+1): 
        for j in range(m+1): 
            if score[i][j]<0: 
                score[i][j]=0
    return score

def display(matrix):
    numRows=len(matrix)
    numCols=len(matrix[0]) #should be equal for all, pick any row
    print(format("-", ">6s"), end="") 
    for ch in seq1: 
        print(format(ch, ">3s"), end="")
    print ("")
    seq2wdash="-"+seq2
    index=0
    for row in range (numRows):
        print (format(seq2wdash[index],"3s"), end="")
        index +=1 
        for col in range(numCols): 
            print(format(matrix[row][col], "3d"), end="")
        print("")

def main2(): 
    print("For the two sequences:", seq1,"and", seq2)
    print("Needleman-Wunsch")
    nwmatrix= nw_fun(seq1,seq2,gap_penalty,match_score,mismatch_score)
    display(nwmatrix)   
    
    print("")
    
    print("Smith-Waterman")
    swmatrix= sw_fun(seq1,seq2,gap_penalty,match_score,mismatch_score)
    display(swmatrix) 

main2()

```

After scores are calculated, the optimum alignment sequence can be found. An example is shown below where many mismatches can be observed.  

![](/SeqAlignmentMtDala.png)


