<!DOCTYPE html>
<html lang="en-US">
  <head>
    <meta charset="utf-8">
    <meta http-equiv="X-UA-Compatible" content="IE=edge">
    <meta name="viewport" content="width=device-width, initial-scale=1">
    <meta name="author" content="Cara Yijin Zou" />
    <meta name="description" content="Describe your website">
    <link rel="shortcut icon" type="image/x-icon" href="/img/favicon.ico">
    <title>Python</title>
    <meta name="generator" content="Hugo 0.61.0" />
    <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
    <link rel="stylesheet" type="text/css" href="/css/main.css" />
    <link rel="stylesheet" type="text/css" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.7.0/css/font-awesome.min.css" />
    <link rel="stylesheet" type="text/css" href="https://fonts.googleapis.com/css?family=Source+Sans+Pro:200,400,200bold,400old" />
    
    <!--[if lt IE 9]>
			<script src="https://oss.maxcdn.com/libs/html5shiv/3.7.0/html5shiv.js"></script>
			<script src="https://oss.maxcdn.com/libs/respond.js/1.4.2/respond.min.js"></script>
		<![endif]-->

    
  </head>

  <body>
    <div id="wrap">

      
      <nav class="navbar navbar-default">
  <div class="container">
    <div class="navbar-header">
      <a class="navbar-brand" href="/"><i class="fa fa-home"></i></a>
      <button type="button" class="navbar-toggle" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
    </div>
    <div class="navbar-collapse collapse" id="navbar">
      <ul class="nav navbar-nav navbar-right">
      
        
        <li><a href="/blog/">BLOG</a></li>
        
        <li><a href="/projects/">PROJECTS</a></li>
        
        <li><a href="/resume/">RESUME</a></li>
        
      
      </ul>
    </div>
  </div>
</nav>

      
      <div class="container">
        <div class="blog-post">
          <h3>
            <strong><a href="/blog/python/">Python</a></strong>
          </h3>
        </div>
        <div class="blog-title">
          <h4>
          December 12, 2019
            &nbsp;&nbsp;
            
          </h4>
        </div>
        <div class="panel panel-default">
          <div class="panel-body">
            <div class="blogpost">
              


<div id="k-mers" class="section level2">
<h2>K-mers</h2>
<p>In addition to R, Python is a valuable tool that can be applied to biology. For example, k-mers are useful in genetics and can be used for a variety of reasons such as sequence assembly, taxonomy, manipulating protein expression, etc. Below is a function that is able to find k-mers in a sequence and record the frequency number in a dictionary.</p>
<pre class="python"><code>
my_seq1 = &quot;CAGCCCAATC&quot;
  
def findkmers(k, seq):
    dict1={}
    a = 0 
    b= k
    k_mer= seq[a:b]
    while len(k_mer)==k: 
        if k_mer in dict1: 
            dict1[k_mer]+= 1
        else: 
            dict1[k_mer]= 1
        a+=1
        b+=1
        k_mer= seq[a:b]
    print (dict1)

def main():
   print(&quot;For the sequence: CAGCCCAATC&quot;)
   for i in range (1,5):
      print(&quot;The&quot;, str(i)+&quot;-mers are :&quot;)
      findkmers(i, my_seq1)

main()</code></pre>
<pre><code>## For the sequence: CAGCCCAATC
## The 1-mers are :
## {&#39;C&#39;: 5, &#39;A&#39;: 3, &#39;G&#39;: 1, &#39;T&#39;: 1}
## The 2-mers are :
## {&#39;CA&#39;: 2, &#39;AG&#39;: 1, &#39;GC&#39;: 1, &#39;CC&#39;: 2, &#39;AA&#39;: 1, &#39;AT&#39;: 1, &#39;TC&#39;: 1}
## The 3-mers are :
## {&#39;CAG&#39;: 1, &#39;AGC&#39;: 1, &#39;GCC&#39;: 1, &#39;CCC&#39;: 1, &#39;CCA&#39;: 1, &#39;CAA&#39;: 1, &#39;AAT&#39;: 1, &#39;ATC&#39;: 1}
## The 4-mers are :
## {&#39;CAGC&#39;: 1, &#39;AGCC&#39;: 1, &#39;GCCC&#39;: 1, &#39;CCCA&#39;: 1, &#39;CCAA&#39;: 1, &#39;CAAT&#39;: 1, &#39;AATC&#39;: 1}</code></pre>
</div>
<div id="sequence-alignment" class="section level2">
<h2>Sequence Alignment</h2>
<p>Sequence alignment is a very important technique used in genetics and is useful for both protein and nucleotide sequences. For example, in my research lab, we use BLAST pairwise alignment on a sequence obtained from DNA sequencing to verify that the DNA matches the known sequence or to check if a desired mutation was successful. Both the Needleman-Wunsch algorithm and Smith-Waterman algorithm are useful ways to find the best alignment.</p>
<p>Here is an example of alignment scores between a sequence in a <em>Mycobacterium tuberculosis</em> sequence compared to humans (low alignment scores) and all organisms (high alignment scores).</p>
<p><img src="/BLASThumanMtDala.png" alt="BLAST Nucleotide MtDala against Human genomic plus transcript (Human G+T)" />
<img src="/BLASTMtDalaOther.png" alt="BLAST Nucleotide MtDala against All Organisms" /></p>
<pre class="python"><code># Use these values to calculate scores
match_score = 2
mismatch_score = -1
gap_penalty = -1

# Make a score matrix with these two sequences
seq1 = &quot;ATGAT&quot;
seq2 = &quot;AAGGAT&quot;

def match_fun(a,b):  # Helper function to define match/mismatch scores
    if a==b:
        score=2 #match score
    else:
        score=-1 #mismatch score
    return score
    
def nw_fun(seq1,seq2,gap_penalty=-1, match_score=1, mismatch_score=-1):
    m=len(seq1) #length of horizontal sequence
    n=len(seq2) #length of vertical sequence
    score =[]
    for row in range(n+2):
        r=[]
        #creates one row
        for col in range(m+2):
            r.append(0)
        score.append(r)
   
    for j in range(0, m + 1): score[0][j] = gap_penalty * j
    for i in range(0, n + 1): score[i][0] = gap_penalty * i
    
    for i in range(1, n + 1): 
        for j in range(1, m + 1):
            insert = score[i - 1][j] + gap_penalty 
            delete = score[i][j - 1] + gap_penalty 
            match = score[i - 1][j - 1] + match_score if seq1[j-1]==seq2[i-1] else score[i-1][j-1]+ mismatch_score
            score[i][j] = max(match, delete, insert)
    return score
    
def sw_fun(seq1,seq2,gap_penalty=-1, match_score=1, mismatch_score=-1):
    m=len(seq1) #length of horizontal sequence
    n=len(seq2) #length of vertical sequence
    score =[]
    for row in range(n+2):
        r=[]
        #creates one row
        for col in range(m+2):
            r.append(0)
        score.append(r)
   
    for j in range(0, m + 1): score[0][j] = gap_penalty * j
    for i in range(0, n + 1): score[i][0] = gap_penalty * i
    
    for i in range(1, n + 1): 
        for j in range(1, m + 1):
            insert = score[i - 1][j] + gap_penalty 
            delete = score[i][j - 1] + gap_penalty 
            match = score[i - 1][j - 1] + match_score if seq1[j-1]==seq2[i-1] else mismatch_score
            score[i][j] = max(match, delete, insert)
            
    for i in range(n+1): 
        for j in range(m+1): 
            if score[i][j]&lt;0: 
                score[i][j]=0
    return score

def display(matrix):
    numRows=len(matrix)
    numCols=len(matrix[0]) #should be equal for all, pick any row
    for row in range (numRows):
        for col in range(numCols): 
            print(format(matrix[row][col], &quot;3d&quot;), end=&quot;&quot;)
        print(&quot;&quot;)

def main2(): 
    print(&quot;For the two sequences:&quot;, seq1,&quot;and&quot;, seq2)
    print(&quot;Needleman-Wunsch&quot;)
    nwmatrix= nw_fun(seq1,seq2,gap_penalty,match_score,mismatch_score)
    display(nwmatrix)   
    
    print(&quot;&quot;)
    
    print(&quot;Smith-Waterman&quot;)
    swmatrix= sw_fun(seq1,seq2,gap_penalty,match_score,mismatch_score)
    display(swmatrix) 

main2()</code></pre>
<pre><code>## For the two sequences: ATGAT and AAGGAT
## Needleman-Wunsch
##   0 -1 -2 -3 -4 -5  0
##  -1  2  1  0 -1 -2  0
##  -2  1  1  0  2  1  0
##  -3  0  0  3  2  1  0
##  -4 -1 -1  2  2  1  0
##  -5 -2 -2  1  4  3  0
##  -6 -3  0  0  3  6  0
##   0  0  0  0  0  0  0
## 
## Smith-Waterman
##   0  0  0  0  0  0  0
##   0  2  1  0  0  0  0
##   0  1  0  0  2  1  0
##   0  0  0  2  1  0  0
##   0  0  0  1  0  0  0
##   0  0  0  0  3  2  0
##   0  0  0  0  2  5  0
##   0  0  0  0  0  0  0</code></pre>
<p>After scores are calculated, the optimum alignment sequence can be found:</p>
<p><img src="/SeqAlignmentMtDala.png" /></p>
</div>

              <hr>
              <div class="related-posts">
                <h5>Related Posts</h5>
                
              </div>
            </div>
          </div>
          <hr>
        <div class="disqus">
  <div id="disqus_thread"></div>
  <script type="text/javascript">

    (function() {
      
      
      if (window.location.hostname == "localhost")
        return;

      var disqus_shortname = '';
      var dsq = document.createElement('script'); dsq.type = 'text/javascript'; dsq.async = true;
      dsq.src = '//' + disqus_shortname + '.disqus.com/embed.js';
      (document.getElementsByTagName('head')[0] || document.getElementsByTagName('body')[0]).appendChild(dsq);
    })();
  </script>
  <noscript>Please enable JavaScript to view the <a href="http://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
  <a href="http://disqus.com/" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</div>
        </div>
      </div>
      
    </div>

    
    <footer>
  <div id="footer">
    <div class="container">
      <p class="text-muted">&copy; All rights reserved. Powered by <a href="https://gohugo.io/">Hugo</a> and
      <a href="http://www.github.com/nurlansu/hugo-sustain/">sustain</a> with â™¥</p>
    </div>
  </div>
</footer>
<div class="footer"></div>


<script src="https://ajax.googleapis.com/ajax/libs/jquery/1.11.3/jquery.min.js"></script>

<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script>
<script src="/js/docs.min.js"></script>
<script src="/js/main.js"></script>

<script src="/js/ie10-viewport-bug-workaround.js"></script>


    
  </body>
</html>
